### ISSUE TITLE
Allow tree views to use icons for symbol kinds on their nodes

### ISSUE THREAD

**Issue by DanTup** [microsoft/vscode#45057]
To display icons in explorer trees we can either pass a path or use `resourceUri`, there's no way to re-use icons used elsewhere in code for symbol types (for example in the workspace symbol searches).

This means that extensions with trees representing code have to ship their own icons (for ex. the "Code Outline" extension has its own: https://github.com/patrys/vscode-code-outline/tree/58008bfa3cc36ad2c18df53b2b4e53dc10a75d55/resources/light) which means they may not be consistent with the icons rendered elsewhere in code.

Is it possible to add something like a `symbolKind: SymbolKind`?

**Comment by jrieken** [comment]
Closing this in favour of #5605 which is something we will do within the next 2 or 3 milestones

**Comment by DanTup** [comment]
@jrieken That sounds good - though in this case this is a specialised tree and not the normal symbol view of my document (it's a sort of render tree we build from static analysis to help the user visualise their UI). With #5606 will it be possible to have a custom tree that can render in a similar way (eg. both a stanadard symbol tree and this custom tree)?

**Comment by DanTup** [comment]
Here's an example of the tree in IntelliJ. The items are not all symbols defined in the document, but in many cases constructor invocations that would create a widget.

<img src="https://lh5.googleusercontent.com/sCxWUKC2gmUnGT-e2R1WYC288cQZhvqbE9Oe7hiKg4cvXN4oDIsUTNmCKWnz2GZOKfF1Ar32Q-Hwg3oZr_DIwo8IrB7MhV7zzFa1pajvWP9RSJj7OpNZnT6z1MbrupAx7FQf_TuF" />

**Comment by jrieken** [comment]
> With #5606 will it be possible to have a custom tree that can render in a similar way (eg. both a stanadard symbol tree and this custom tree)?

Yeah, should be possible... The idea is to extend the DocumentSymbelProvider so that it can return hierarchical data. However, the idea is also to allow for different representations of that data, e.g. a break-crumb and/or a tree. Also likely that the tree will sit where the minimap sits. 

What you are trying to achieve seems a little further away from "normal" outline, tho there is no normal outline...

**Comment by DanTup** [comment]
@jrieken Yeah, I understand our need is somewhat unusual. Using the DocumentSymbolProvider for this would feel a bit weird since it's not semantically the same.

Another possible option might be to provide us with a way to get the `Uri` for a `Uri` fo built-in icons that we could provide to the existing explorer tree? That would be fairly generic/reusable and probably relatively simple to implement?

**Comment by jrieken** [comment]
cc @sandy081 who is doing some work on icons

**Comment by sandy081** [comment]
@jrieken Are there constants in the API  representing document symbol icons?

**Comment by DanTup** [comment]
I don't know if this answers the question but there's an enum named `SymbolKind` which is used in the `SymbolInformation` types returned by the symbol providers. My need is to be able to display an icon for one of them in the contributed tree.

**Comment by sandy081** [comment]
@DanTup Thanks.

@jrieken How about extending `iconPath` to take `SymbolKind` constant in addition ?

**Comment by jrieken** [comment]
> How about extending iconPath to take SymbolKind constant in addition 

That doesn't sound right to me... I think we should make them part of the themes and, assuming my understanding of themes is correct, make them reusablable by that

**Comment by sandy081** [comment]
Ok, Makes sense.

**Comment by axetroy** [comment]
agree.

**Comment by DanTup** [comment]
Is it likely there may be any progress on this in the near future? I'm currently thinking about just taking a copy of the icons from [here](https://github.com/microsoft/vscode/tree/a6ee65e647e6eeb1b6926d4100276e7afff14510/src/vs/editor/contrib/documentSymbols/media) and periodically updating them. It doesn't feel like an ideal solution, but users have been asking for a Flutter Outline for well over a year and I'd really like to deliver *something*.

**Comment by sandy081** [comment]
No progress.. also given that outline view is provided out of the box, not sure if it is needed for other extensions.

**Comment by DanTup** [comment]
I tried many times to use the built-in outline, but it does support being extended as required here and I was guided towards a custom tree. Specifically, it does not have:

- Ability to have custom icons (https://github.com/microsoft/vscode/issues/51334 closed as a dupe of https://github.com/microsoft/vscode/issues/50186 then rejected)
- Ability to have tree nodes that don't also show up in the Document Symbol list (because without the tree context they are ambiguous - see https://github.com/microsoft/vscode/issues/51332#issuecomment-503501618).

I'm also not sure it has selected events, which I'd also need to set contexts correctly so my context menus can work.

I've got prototypes going both ways, but neither are great. I would prefer to extend the built-in outline, but more requests have been rejected for supporting that than the custom tree, so I figure it has to be this way.
